### PR TITLE
Add MATLAB Bindings

### LINKED ISSUE #102: Implement a Matlab API
There is an idea to implement a preCICE API for Matlab. For this, we would need to call C/C++ functions from Matlab.

I found that the [coder.ceval](https://de.mathworks.com/help/simulink/slref/coder.ceval.html) can do this, I am not very sure about the implementation though.

Until now there has been at least one potential user that wants to couple their Matlab structure solver with preCICE.

### PR DESCRIPTION
**Pull request by BenjaminRueth**
Closes #102.

This branch currently contains all the changes by @Dominanz that were necessary for his thesis. One particularly useful contribution for preCICE are the MATLAB bindings that have been developed by @Dominanz. We want to merge @Dominanz implementation of the MATLAB bindings into https://github.com/precice/precice within this PR.

I would suggest to release the MATLAB bindings as soon as possible in order to be able to provide them to our users. However, I would suggest to explicitly declare them as **experimental** since I expect that we will need to make some experiences until we arrive at the final version of the Interface.

# Requirements

We require the following components for the MATLAB bindings:

* all source code that is necessary for installation of the MATLAB bindings
* documentation, such that a user can install the MATLAB bindings
* a solverdummy using the MATLAB bindings

# Todo

Before we can merge this PR, we have to do the following things:

- [x] get rid of everything that is not directly connected to the MATLAB bindings (see requirements above)
- [x] add documentation, if it is missing (-> documentation is in good shape, more will follow in a later PR)
- [x] write tests for the MATLAB bindings (might also be postponed after merging, since we might need to test the MATLAB bindings using a system test on https://github.com/precice/systemtests/tree/master/Test_bindings)
- [x] update the [Changelog](https://github.com/precice/precice/blob/develop/CHANGELOG.md)
- [x] update the [wiki](https://github.com/precice/precice/wiki/Non%E2%80%93standard-APIs) as soon as PR is merged

### CODE REVIEW

File: src/precice/bindings/matlab/+precice/@Constants/private/preciceGateway.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+class MexFunction: public matlab::mex::Function {
+private:
+    ArrayFactory factory;
+    void myMexPrint(const std::string text) {
+        std::cout << "MEX constants gateway: " << text << std::endl;
+    }
+
+public:
+    MexFunction(): factory{} {}
```
Comment (fsimonis) [inline src/precice/bindings/matlab/+precice/@Constants/private/preciceGateway.cpp]: The compiler can auto generate this constructor.
```suggestion
```

File: src/precice/bindings/matlab/+precice/@Constants/private/preciceGateway.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    void operator()(ArgumentList outputs, ArgumentList inputs) {
+        // define the constantID
+        TypedArray<uint8_t> functionIDArray = inputs[0];
+        int constantID = functionIDArray[0];
```
Comment (fsimonis) [inline src/precice/bindings/matlab/+precice/@Constants/private/preciceGateway.cpp]: Can't you merge these two calls?

File: src/precice/bindings/matlab/+precice/@Constants/private/preciceGateway.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#include "precice/SolverInterface.hpp"
+
+using namespace matlab::data;
+using matlab::mex::ArgumentList;
+using namespace precice::constants;
+
+class MexFunction: public matlab::mex::Function {
+private:
+    ArrayFactory factory;
+    void myMexPrint(const std::string text) {
```
Comment (fsimonis) [inline src/precice/bindings/matlab/+precice/@Constants/private/preciceGateway.cpp]: `myMexPrint` is private and used once. So let's just put it into the below function.

File: src/precice/bindings/matlab/+precice/@Constants/private/preciceGateway.cpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                result = actionWriteInitialData();
+                break;
+            case 1:
+                result = actionWriteIterationCheckpoint();
+                break;
+            case 2:
+                result = actionReadIterationCheckpoint();
+                break;
+            default:
+                myMexPrint("An unknown ID was passed.");
```
Comment (fsimonis) [inline src/precice/bindings/matlab/+precice/@Constants/private/preciceGateway.cpp]: Wouldn't the use of `std::cerr` more appropriate here?

File: tools/solverdummies/precice-config.xml [hunk 5]
```diff
@@ -3,7 +3,7 @@
 <precice-configuration>
 
   <log>
-    <sink type="stream" output="stdout"  filter= "%Severity% > debug" format="preCICE:%ColorizedSeverity% %Message%" enabled="true" />
```
Comment (fsimonis) [inline tools/solverdummies/precice-config.xml]: Unrelated change

File: src/precice/impl/SolverInterfaceImpl.cpp [hunk 6]
```diff
@@ -100,6 +100,10 @@ void SolverInterfaceImpl:: configure
 (
   const std::string& configurationFileName )
 {
+  mesh::Mesh::resetGeometryIDsGlobally();
```
Comment (fsimonis) [inline src/precice/impl/SolverInterfaceImpl.cpp]: @uekerman Can you check your fix?

File: tools/solverdummies/matlab-legacy/solverdummy.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+ * To run, run the following command in MATLAB:
+ * solverdummy configFile solverName meshName
+ * OR, equivalently,
+ * solverdummy('configFile','solverName','meshName')
+ * since MATLAB syntax works like this.
+*/
+
+#include "precice/SolverInterface.hpp"
+#include "precice/Constants.hpp"
+#include <math.h>
```
Comment (fsimonis) [inline tools/solverdummies/matlab-legacy/solverdummy.cpp]:
```suggestion
#include <cmath>
```

File: tools/solverdummies/matlab-legacy/solverdummy.cpp [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+#include <matrix.h>
+#include <mex.h>
+
+void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
+{
+  mexPrintf("Hello World!\n");
```
Comment (fsimonis) [inline tools/solverdummies/matlab-legacy/solverdummy.cpp]:
```suggestion
```

File: tools/solverdummies/matlab-legacy/solverdummy.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  std::string solverName(mxArrayToString(prhs[1]));
+  std::string meshName(mxArrayToString(prhs[2]));
+  int N = 1;
+  SolverInterface interface(solverName, commRank, commSize);
+  interface.configure(configFileName);
+
+  int meshID = interface.getMeshID(meshName);
+  int dimensions = interface.getDimensions();
+  mexPrintf(meshName.c_str());
+  std::vector<std::vector<double>> vertices(N, std::vector<double>(dimensions, 0));
```
Comment (fsimonis) [inline tools/solverdummies/matlab-legacy/solverdummy.cpp]: We should populate the dimensions somehow.
`std::generate` and some random distribution maybe?

File: src/precice/bindings/matlab/+precice/@SolverInterface/private/preciceGateway.cpp [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+    }
+    
+public:
+    MexFunction(): constructed{false}, factory{}, interface{NULL} {}
+
+    void operator()(ArgumentList outputs, ArgumentList inputs) {
+        bool valid=false;
+        // define the functionID
+        const TypedArray<uint8_t> functionIDArray = inputs[0];
+        int functionID = functionIDArray[0];
```
Comment (fsimonis) [inline src/precice/bindings/matlab/+precice/@SolverInterface/private/preciceGateway.cpp]: Can't we merge these two calls?

File: src/precice/bindings/matlab/+precice/@SolverInterface/private/preciceGateway.cpp [hunk 11]
```diff
@@ -0,0 +1,397 @@
...
+        if (functionID==0 && constructed) {
+            myMexPrint("Constructor was called but interface is alread construced.");
+            return;
+        }
+        if (!constructed && functionID!=0) {
+            myMexPrint("Interface was not construced before.");
+            return;
+        }
+        
+        switch (functionID) {
```
Comment (fsimonis) [inline src/precice/bindings/matlab/+precice/@SolverInterface/private/preciceGateway.cpp]: Put the function ids in a `enum class FunctionID { Constructor, Destructor, ... } `.
Then you can `static_cast<FunctionID>(functionID)` and use that in the `switch(...)`.
This makes the following `case(...)` statements more explicit and less error prone: `case FunctionID::Constructor`.

Use `return` in the cases instead of `break` as there is no error handling.

File: src/precice/impl/SolverInterfaceImpl.cpp (hunk 6 above)
Comment (uekerman) [reply src/precice/impl/SolverInterfaceImpl.cpp]: This is a workaround that I suggested to @Dominanz, see #378 
We should exclude this from the PR as it is unrelated to the Matlab bindings.

**Comment by BenjaminRueth** [comment]
From my perspective unrealistic to get this done until the release. Let's postpone it to the next one.

File: src/precice/impl/SolverInterfaceImpl.cpp (hunk 6 above)
Comment (BenjaminRueth) [reply src/precice/impl/SolverInterfaceImpl.cpp]: I agree. Does this mean that we do not want to introduce this change at all or just introduce it in a different PR?

File: src/precice/SolverInterface.hpp [hunk 12]
```diff
@@ -697,9 +697,9 @@ class SolverInterface
    * @see SolverInterface::setMeshVertex()
    */
   void writeScalarData (
-    int    dataID,
-    int    valueIndex,
-    double value );
+    int           dataID,
+    int           valueIndex,
+    const double& value );
```
Comment (BenjaminRueth) [inline src/precice/SolverInterface.hpp]: These changes should not be part of this PR. Similar to https://github.com/precice/precice/pull/494/files#r320836465

File: src/precice/SolverInterface.cpp [hunk 13]
```diff
@@ -290,9 +290,9 @@ void SolverInterface:: writeBlockScalarData
 
 void SolverInterface:: writeScalarData
 (
-  int    dataID,
-  int    valueIndex,
-  double value )
+  int           dataID,
+  int           valueIndex,
+  const double& value )
```
Comment (BenjaminRueth) [inline src/precice/SolverInterface.cpp]: These changes should not be part of this PR. Similar to https://github.com/precice/precice/pull/494/files#r320836465

File: src/precice/SolverInterface.hpp (hunk 12 above)
Comment (BenjaminRueth) [reply src/precice/SolverInterface.hpp]: I created a PR for adding these changes independently: https://github.com/precice/precice/pull/530

File: src/precice/SolverInterface.cpp (hunk 13 above)
Comment (BenjaminRueth) [reply src/precice/SolverInterface.cpp]: I created a PR for adding these changes independently: https://github.com/precice/precice/pull/530

File: src/precice/SolverInterface.hpp (hunk 12 above)
Comment (BenjaminRueth) [reply src/precice/SolverInterface.hpp]: Changes are now removed from this PR.

File: src/precice/SolverInterface.cpp (hunk 13 above)
Comment (BenjaminRueth) [reply src/precice/SolverInterface.cpp]: 0615ee2 removes these changes from this PR.

File: src/precice/SolverInterface.hpp (hunk 12 above)
Comment (BenjaminRueth) [reply src/precice/SolverInterface.hpp]: See 0615ee2.

File: src/precice/impl/SolverInterfaceImpl.cpp (hunk 6 above)
Comment (BenjaminRueth) [reply src/precice/impl/SolverInterfaceImpl.cpp]: I removed the code changes from this PR (https://github.com/precice/precice/pull/494/commits/e922868e6258bb35cf095106483e9307fc679c56) and added a comment in the `README.md` file of the bindings (40b16c7).

**Comment by BenjaminRueth** [comment]
> I just cloned the PR branch after merging develop, built preCICE (to work with MPICH), compiled the MATLAB bindings and the solverdummy works without a problem.

That's good news. 

As soon as @fsimonis has aproved the PR, we can merge it. There are some comments in his review that we should either directly address in this PR or at least transfer to issues, such that we do not forget.

**Comment by MakisH** [comment]
This pull request has been mentioned on **preCICE**. There might be relevant details there:

https://precice.discourse.group/t/i-was-asked-to-try-a-pull-request-how-can-i-do-that/38/2

**Comment by BenjaminRueth** [comment]
I would like to merge this PR into a new feature branch `precice:draft_MATLAB_bindings` on the preCICE repository, now. Some issues still remain open end will be solved by another PR, before we merge everything into `precice:develop`. Using this approach, we can clearly identify @Dominanz contribution to the code and let @gilbertolem add his contributions on his own fork.

From my point of view the following issues still remain open:

* put everything into `extras/bindings` see https://github.com/precice/precice/pull/494#pullrequestreview-283549411
* use enums for a clearer mapping of cpp to matlab API functions in the [`preciceGateway`](https://github.com/Dominanz/precice/blob/b3d7ede68059a39f9daef673bdb3661c8910f205/src/precice/bindings/matlab/%2Bprecice/%40SolverInterface/SolverInterface.m#L47-L49)

@fsimonis do you agree? If yes, you can just merge this PR from my point of view.
